### PR TITLE
Update DeepStream and DeepStream Python bindings

### PR DESCRIPTION
**Pull request by ichergui**

### CODE REVIEW

File: recipes-devtools/deepstream/deepstream-6.1_6.1.0-1.bb [hunk 1]
```diff
@@ -82,6 +89,32 @@ do_install() {
     cp -R --preserve=mode,timestamps ${S}${DEEPSTREAM_PATH}/sources/includes/* ${D}${includedir}/
 
     cp -R --preserve=mode,timestamps ${S}${DEEPSTREAM_PATH}/sources/ ${D}${DEEPSTREAM_PATH}/
+
+    # XXX---
+    # Some of the binaries are not using the right SONAME
+    # in its DT_NEEDED ELF header, so we have to rewrite it to prevent
+    # a broken runtime dependency.
```
Comment (madisongh) [inline recipes-devtools/deepstream/deepstream-6.1_6.1.0-1.bb:96]: This is a bit concerning, as substituting a different version of a library (particularly C++) could result in strange runtime errors.

Looks like most of these are for the `libyaml-cpp` version difference.  Maybe we could instead add a recipe for the older version (that could co-reside with the upstream version, not replace) and add that as a dependency?  We've done that for other cases like this.

Comment (ichergui) [reply recipes-devtools/deepstream/deepstream-6.1_6.1.0-1.bb:96]: Yes, I can do that 
I'm doing some tests and I will share here

File: recipes-devtools/deepstream/deepstream-6.1_6.1.0-1.bb [hunk 2]
```diff
@@ -82,6 +89,32 @@ do_install() {
...
+    patchelf --replace-needed libyaml-cpp.so.0.6 libyaml-cpp.so.0.7 ${D}${DEEPSTREAM_PATH}/lib/libpostprocess_impl.so
+    patchelf --replace-needed libyaml-cpp.so.0.6 libyaml-cpp.so.0.7 ${D}${DEEPSTREAM_PATH}/lib/libnvds_yml_parser.so
+    patchelf --replace-needed libyaml-cpp.so.0.6 libyaml-cpp.so.0.7 ${D}${DEEPSTREAM_PATH}/lib/libnvds_3d_depth2point_datafilter.so
+    patchelf --replace-needed libyaml-cpp.so.0.6 libyaml-cpp.so.0.7 ${D}${DEEPSTREAM_PATH}/lib/libnvds_3d_depth_datasource.so
+    patchelf --replace-needed libyaml-cpp.so.0.6 libyaml-cpp.so.0.7 ${D}${DEEPSTREAM_PATH}/lib/libnvds_3d_gl_datarender.so
+    patchelf --replace-needed libyaml-cpp.so.0.6 libyaml-cpp.so.0.7 ${D}${DEEPSTREAM_PATH}/lib/libnvds_msgconv.so.1.0.0
+    patchelf --replace-needed libyaml-cpp.so.0.6 libyaml-cpp.so.0.7 ${D}${DEEPSTREAM_PATH}/lib/libnvds_nvmultiobjecttracker.so
+    patchelf --replace-needed libdns_sd.so.1.0.0 libdns_sd.so.1 ${D}${DEEPSTREAM_PATH}/lib/libnvds_nvmultiobjecttracker.so
+    patchelf --replace-needed libdns_sd.so.1.0.0 libdns_sd.so.1 ${D}${DEEPSTREAM_PATH}/lib/libnvds_nmos.so
```
Comment (madisongh) [inline recipes-devtools/deepstream/deepstream-6.1_6.1.0-1.bb:113]: These for `libdns_sd` look strange.  Probably OK, as I can't think of a better workaround.

File: recipes-devtools/deepstream/deepstream-6.1_6.1.0-1.bb [hunk 3]
```diff
@@ -82,6 +89,32 @@ do_install() {
...
+    patchelf --replace-needed libcufft.so libcufft.so.10 ${D}${DEEPSTREAM_PATH}/lib/libnvds_nvmultiobjecttracker.so
+    patchelf --replace-needed libcublas.so libcublas.so.11 ${D}${DEEPSTREAM_PATH}/lib/libnvds_nvmultiobjecttracker.so
+    patchelf --replace-needed libcufft.so libcufft.so.10 ${D}${DEEPSTREAM_PATH}/lib/libnvds_audiotransform.so
+    # ---XXX
```
Comment (madisongh) [inline recipes-devtools/deepstream/deepstream-6.1_6.1.0-1.bb:117]: We could solve these by installing the `.so` symlinks, as we have done in other cases where NVIDIA has made this same rookie mistake.

**Review by madisongh — COMMENTED** [review]
I'd be OK with this going in as-is, but a better solution for most of the `patchelf` hacks will be needed at some point.

File: recipes-devtools/deepstream/deepstream-6.1_6.1.0-1.bb (hunk 3 above)
Comment (ichergui) [reply recipes-devtools/deepstream/deepstream-6.1_6.1.0-1.bb:117]: In this case, we need `-dev` packages right ?

Comment (madisongh) [reply recipes-devtools/deepstream/deepstream-6.1_6.1.0-1.bb:117]: We could put the symlink into a separate package, like we did for libvisionworks.

Comment (ichergui) [reply recipes-devtools/deepstream/deepstream-6.1_6.1.0-1.bb:117]: I will try that 
Thanks

**Comment by dwalkes** [comment]
@ichergui I'm as far as building into an image but haven't tried runnning. I see you said you have some tests in progress, let me know if there's anything specific you'd like me to try.  Should we add some tests in https://docs.google.com/spreadsheets/d/1EWG5QZvtWdk74gz48D4xCbegK-bsqMFJ8XOsyGdUUeI/edit#gid=145133208 ?  I was thinking about updating that sheet this week with the latest wip branch.

**Comment by ichergui** [comment]
Thanks @dwalkes 
I will update my PR today and we need more tests to be done 
I'm trying to push bash script to automate the testing as we have done with VPI, TRT, GST, ...

**Comment by ichergui** [comment]
I tried the `deepstream` sample applications and they are working fine.
Please check the log files for reference.

* Test1 
```
$ cd /opt/nvidia/deepstream/deepstream-6.1/sources/apps/sample_apps/deepstream-test1/
$ deepstream-test1-app /opt/nvidia/deepstream/deepstream-6.1/samples/streams/sample_720p.h264
```
[deepstream-test1-app.log](https://github.com/OE4T/meta-tegra-community/files/8869460/deepstream-test1-app.log)

* Test2
```
$ cd /opt/nvidia/deepstream/deepstream-6.1/sources/apps/sample_apps/deepstream-test2/
$ deepstream-test2-app /opt/nvidia/deepstream/deepstream-6.1/samples/streams/sample_720p.h264
```

[deepstream-test2-app.log](https://github.com/OE4T/meta-tegra-community/files/8869463/deepstream-test2-app.log)


* Test3 
```
$ cd /opt/nvidia/deepstream/deepstream-6.1/sources/apps/sample_apps/deepstream-test3/
$ deepstream-test3-app file:///opt/nvidia/deepstream/deepstream-6.1/samples/streams/sample_1080p_h264.mp4 file:///opt/nvidia/deepstream/deepstream-6.1/samples/streams/sample_720p.mp4

```
[deepstream-test3-app.log](https://github.com/OE4T/meta-tegra-community/files/8869468/deepstream-test3-app.log)

**Comment by ichergui** [comment]
I tried the `deepstream` python sample applications and they are working fine as well.
Please check the log files for reference.

* Test1 
```
$ cd /opt/nvidia/deepstream/deepstream-6.1/sources/deepstream_python_apps/apps/deepstream-test1
$ python3 deepstream_test_1.py /opt/nvidia/deepstream/deepstream-6.1/samples/streams/sample_720p.h264
```
[python3-deepstream-test1-app.log](https://github.com/OE4T/meta-tegra-community/files/8869897/python3-deepstream-test1-app.log)


* Test2
```
$ cd /opt/nvidia/deepstream/deepstream-6.1/sources/deepstream_python_apps/apps/deepstream-test2
$ python3 deepstream_test_2.py /opt/nvidia/deepstream/deepstream-6.1//samples/streams/sample_720p.h264
```

[python3-deepstream-test2-app.log](https://github.com/OE4T/meta-tegra-community/files/8869900/python3-deepstream-test2-app.log)


* Test3
```
$ cd /opt/nvidia/deepstream/deepstream-6.1/sources/deepstream_python_apps/apps/deepstream-test3
$ python3 deepstream_test_3.py -i file:///opt/nvidia/deepstream/deepstream-6.1/samples/streams/sample_1080p_h264.mp4 file:///opt/nvidia/deepstream/deepstream-6.1/samples/streams/sample_720p.mp4

```

[python3-deepstream-test3-app.log](https://github.com/OE4T/meta-tegra-community/files/8869901/python3-deepstream-test3-app.log)

**Comment by ichergui** [comment]
Tests were done on `Jetson Xavier AGX` 
Here is the result of the test:

```
Frame Number= 1441 Number of Objects= 15 Vehicle_count= 12 Person_count= 3
Frame Number= 1441 Number of Objects= 0 Vehicle_count= 0 Person_count= 0
Frame Number= 1442 Number of Objects= 0 Vehicle_count= 0 Person_count= 0
End-of-stream
Exiting app

=== PASS:  py_deepstream_test3 ===
Tests run:     6
Tests passed:  6
Tests skipped: 0
Tests failed:  0

root@jetson-agx-xavier-devkit:~#
```

**Review by madisongh — APPROVED** [review]
LGTM

**Pull request merged**